### PR TITLE
fix(anvil): proper eip-1559 base fee implementation

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
Closes ##1919

## Solution
Contains various fixes:
* update next base fee when new block mined
* make gas price exclusive pre London
* use fork fee history if forked block requested in `eth_feeHistory`
* reset base fee on fork reset
* validate tx's fee is higher than current base fee

### CODE REVIEW

**Comment by mattsse** [comment]
@joshieDo I've no idea what the failing test does.

could you take a look whether this is a regression or if it needs to be updated somehow?

File: anvil/src/eth/fees.rs [hunk 1]
```diff
@@ -71,9 +99,49 @@ impl FeeManager {
...
+pub fn calc_next_base_fee(
+    gas_used: U256,
+    gas_limit: U256,
+    last_fee_per_gas: U256,
+    elasticity: f64,
+) -> u64 {
+    let gas_used = gas_used.as_u64() as f64;
+    let gas_limit = gas_limit.as_u64() as f64;
+    let gas_target = gas_limit / elasticity;
+    let last_gas_used = gas_used / (gas_target * elasticity);
```
Comment (gakonst) [inline anvil/src/eth/fees.rs:128]: can this have a rounding error?

File: anvil/src/eth/fees.rs [hunk 2]
```diff
@@ -71,9 +99,49 @@ impl FeeManager {
...
+        (last_fee_per_gas + (last_fee_per_gas * increase)) as u64
+    } else if last_gas_used < 0.5 {
+        // decrease gas
+        let increase = ((0.5 - last_gas_used) * 2f64) * elasticity;
+
+        (last_fee_per_gas - (last_fee_per_gas * increase)) as u64
+    } else {
+        // same base gas
+        last_fee_per_gas as u64
+    }
```
Comment (gakonst) [inline anvil/src/eth/fees.rs:144]: this must clip at 12.5% max, is this the case? is that what `elasticity` does?

Comment (mattsse) [reply anvil/src/eth/fees.rs:144]: yeh elasticity is `1/8`(12.5%)

File: anvil/src/eth/fees.rs (hunk 1 above)
Comment (mattsse) [reply anvil/src/eth/fees.rs:128]: possible, but negligible I think

but could replace this with a decimal type in a followup?

**Review by gakonst — APPROVED** [review]
sgtm

**Pull request merged**